### PR TITLE
Dns over http2 5773 v14

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5773

Describe changes:
- analyze DNS over HTTP2

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1734

https://github.com/OISF/suricata/pull/11369 with greener CI.
@victorjulien should I squash the 2 last commits in ?
They are are help to help you review your last comments...

Draft because of TODO : 
- Fix dns json schema and logging first (that is https://redmine.openinfosecfoundation.org/issues/6281 )
- Use config for dns logging...

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_asan.
  

Pipeline 21273

**Comment by victorjulien** [comment]
Passed my QA. Ran this PR with SV PR https://github.com/OISF/suricata-verify/pull/1734. Local pipeline 4966, run 511.

File: src/detect.c [hunk 1]
```diff
@@ -1130,10 +1130,14 @@ static bool DetectRunTxInspectRule(ThreadVars *tv,
         {
             void *tx_ptr = DetectGetInnerTx(tx->tx_ptr, f->alproto, engine->alproto, flow_flags);
             if (tx_ptr == NULL) {
-                /* special case: file_data on 'alert tcp' will have engines
-                 * in the list that are not for us. */
-                engine = engine->next;
-                continue;
+                if (engine->alproto != 0) {
```
Comment (victorjulien) [inline src/detect.c:1133]: use ALPROTO_UNKNOWN here?

File: rust/src/http2/http2.rs [hunk 2]
```diff
@@ -124,6 +128,15 @@ pub struct HTTP2Frame {
     pub data: HTTP2FrameTypeData,
 }
 
+#[derive(Debug, Default)]
+pub struct DohHttp2Tx {
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:132]: can you add field description comments, esp on the index to these 2 fields

File: rust/src/http2/logger.rs [hunk 3]
```diff
@@ -278,6 +282,35 @@ fn log_http2(tx: &HTTP2Transaction, js: &mut JsonBuilder) -> Result<bool, JsonEr
...
 
+    if let Some(doh) = &tx.doh {
+        js.open_object("dns")?;
+        if let Some(dtx) = &doh.dns_request_tx {
+            let mark = js.get_mark();
+            let mut has_dns_query = false;
+            js.open_array("query")?;
+            for i in 0..0xFFFF {
+                let mut jsa = JsonBuilder::try_new_object()?;
+                if !SCDnsLogJsonQuery(dtx, i, 0xFFFFFFFFFFFFFFFF, &mut jsa) {
```
Comment (victorjulien) [inline rust/src/http2/logger.rs:293]: can we use a named constant for this?

File: src/app-layer-detect-proto.c [hunk 4]
```diff
@@ -1844,6 +1844,16 @@ bool AppLayerRequestProtocolTLSUpgrade(Flow *f)
     return AppLayerRequestProtocolChange(f, 443, ALPROTO_TLS);
 }
 
+void AppLayerForceProtocolChange(Flow *f, AppProto new_proto)
```
Comment (victorjulien) [inline src/app-layer-detect-proto.c:1847]: can this get a comment explaining the logic of only setting f->alproto to new_proto?

File: src/app-layer-protos.h [hunk 5]
```diff
@@ -101,6 +109,48 @@ static inline bool AppProtoEquals(AppProto sigproto, AppProto alproto)
     return false;
 }
 
+// whether a signature AppProto matches a flow (or signature) AppProto
```
Comment (victorjulien) [inline src/app-layer-protos.h:112]: can you add a comment here on the interplay between doh2 and http2 and sigs using either?

File: src/detect.c [hunk 6]
```diff
@@ -1050,6 +1050,24 @@ DetectRunTxSortHelper(const void *a, const void *b)
 #define TRACE_SID_TXS(sid,txs,...)
 #endif
 
+// Get inner transaction for engine
+void *DetectGetInnerTx(void *tx_ptr, AppProto alproto, AppProto engine_alproto, uint8_t flow_flags)
+{
+    if (alproto == ALPROTO_DOH2) {
```
Comment (victorjulien) [inline src/detect.c:1056]: is this expected to be the common case for calls to this function? Or should we put the non-doh2 first to improve branch prediction?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Nice work so far. Some comments inline.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21346

File: src/detect.c (hunk 1 above)
Comment (catenacyber) [reply src/detect.c:1133]: ok (is written 0 in master)

File: rust/src/http2/http2.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:132]: ok

File: src/app-layer-detect-proto.c (hunk 4 above)
Comment (catenacyber) [reply src/app-layer-detect-proto.c:1847]: alproto_ts and alproto_tc are also set to new_proto

File: src/app-layer-protos.h (hunk 5 above)
Comment (catenacyber) [reply src/app-layer-protos.h:112]: ok

File: src/app-layer-detect-proto.c (hunk 4 above)
Comment (catenacyber) [reply src/app-layer-detect-proto.c:1847]: adding a comment to the function anyways

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11428

**Pull request closed without merging**